### PR TITLE
WIP Add the recompute_grad decorator from tf.contrib

### PR DESCRIPTION
**Pull request by olesalscheider**
Fix https://github.com/tensorflow/addons/issues/218

I tried to port the [recompute_grad](https://www.tensorflow.org/versions/r1.12/api_docs/python/tf/contrib/layers/recompute_grad?hl=en) decorator to TF 2.0.
However, there are still three TODOs in the code where I don't know how to handle them.
Also, the tests still fail in graph mode...

### CODE REVIEW

File: tensorflow_addons/layers/decorators_test.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    with tf.GradientTape() as tape:
+      x = tf.ones((2, 4), tf.float32)
+      y = layer_with_recompute(x)
+      loss = tf.math.reduce_sum(y)
+    tvars = dense.trainable_variables + \
+            batch_normalization.trainable_variables
+    tape.gradient(loss, tvars)
+
+    # TODO: What to do with this?
+    #update_ops = ops.get_collection(ops.GraphKeys.UPDATE_OPS)
```
Comment (facaiy) [inline tensorflow_addons/layers/decorators_test.py]: Hi, Niels, I'm not sure what do you mean here. If we want to collect update_ops, perhaps we could build [tf.keras.Sequential](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/Sequential) or create a subclass of [tf.keras.Model](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/Model). And then collect all update_ops by `[layer.updates for layer in model.layers]`, does it work for you?

Comment (facaiy) [reply tensorflow_addons/layers/decorators_test.py]: You might want to read: https://www.tensorflow.org/alpha/guide/keras/functional

Comment (olesalscheider) [reply tensorflow_addons/layers/decorators_test.py]: I mean what this check is supposed to do the following: Because of the recompute_grad decorator the function is evaluated multiple times. But we want to only update the batch normalization statistics on the first run, because otherwise we get wrong statistics.
So far in graph mode the batch normalization updates were added to the UPDATE_OPS collection. Therefore the code above could remove the duplicate update ops before they were carried out.

But now in eager mode the updates are carried out in place. So we can't take them back anymore.
Therefore we need a way to avoid that the update happens in the first place in eager mode.
I don't know how to do that without modifying the BatchNormalization layer so this still is a TODO.

Comment (facaiy) [reply tensorflow_addons/layers/decorators_test.py]: Got it, thank you for the clarification, Niels. 

I'll be out of town in the next two weeks, @WindQAQ @kyleabeauchamp Tzu-Wei, Kyle, could you take a look and help Niels out when time allows? Thank you :-)

Comment (facaiy) [reply tensorflow_addons/layers/decorators_test.py]: @pavithrasv @qlzh727 @mihaimaruseac Hi, Pavithra, Qianli, Mihai, do you have any good idea to control the updating behavior of layer? Thank you very much, cc @fchollet

**Comment by seanpmorgan** [comment]
Again thank you very much for your work here... apologies for the miscommunication w.r.t this being added to tf core. Its on the agenda for today's meeting to discuss.
